Slim/Control/Queries.pm

```diff
@@ -1882,6 +1882,147 @@ sub genresQuery {
+	my $tags          = $request->getParam('tags') || '';
+
+	my $sql  = 'SELECT %s FROM contributors ';
+	my $w    = [];
+	my $p    = [];
+
+	# Manage joins
+	if (defined $trackID) {
+		$sql .= 'JOIN contributor_track ON contributors.id = contributor_track.contributor ';
+		push @{$w}, 'contributor_track.track = ?';
+		push @{$p}, $trackID;
+	}
+	else {
+		# ignore these if we have a track.
+		$sql .= 'JOIN contributor_album ON contributors.id = contributor_album.contributor ';
+
+		if (defined $contributorID) {
+
+			# handle the case where we're asked for the VA id => return compilations
+			if ($contributorID == Slim::Schema->variousArtistsObject->id) {
+				$sql .= 'JOIN albums ON contributor_album.album = albums.id ';
+				push @{$w}, 'albums.compilation = ?';
+				push @{$p}, 1;
+			}
+			else {
+				push @{$w}, 'contributor_album.contributor = ?';
+				push @{$p}, $contributorID;
+			}
+		}
+
+		if ( $libraryID ) {
+			$sql .= 'JOIN library_contributor ON library_contributor.contributor = contributors.id ';
+			push @{$w}, 'library_contributor.library = ?';
+			push @{$p}, $libraryID;
+		}
+
+		if (defined $albumID) {
+			push @{$w}, 'contributor_album.album = ?';
+			push @{$p}, $albumID;
+		}
+
+		if (defined $year || defined $workID) {
+			$sql .= 'JOIN contributor_track ON contributors.id = contributor_track.contributor ';
+			$sql .= 'JOIN tracks ON tracks.id = contributor_track.track ';
+
+			if (defined $year) {
+				push @{$w}, 'tracks.year = ?';
+				push @{$p}, $year;
+			}
+			if (defined $workID) {
+				if ( $workID eq "-1" ) {
+					push @{$w}, 'tracks.work IS NOT NULL';
+				} else {
+					push @{$w}, 'tracks.work = ?';
+					push @{$p}, $workID;
+				}
+			}
+		}
+	}
+
+	if ( @{$w} ) {
+		$sql .= 'WHERE ';
+		my $s = join( ' AND ', @{$w} );
+		$s =~ s/\%/\%\%/g;
+		$sql .= $s . ' ';
+	}
+
+	my $dbh = Slim::Schema->dbh;
+
+	if (defined $trackID) {
+		$sql = sprintf($sql, 'GROUP_CONCAT(DISTINCT contributor_track.role)');
+	} else {
+		$sql = sprintf($sql, 'GROUP_CONCAT(DISTINCT contributor_album.role)');
+	}
+
+	my $stillScanning = Slim::Music::Import->stillScanning();
+
+	# now build the result
+
+	if ($stillScanning) {
+		$request->addResult('rescan', 1);
+	}
+
+	my $cacheKey = md5_hex($sql . join( '', @{$p} ) . Slim::Music::VirtualLibraries->getLibraryIdForClient($client));
+
+	my $roles = $cache->{$cacheKey};
+
+	if ( !$roles ) {
+		# Limit the real query
+		if ( $index =~ /^\d+$/ && $quantity =~ /^\d+$/ ) {
+			$sql .= "LIMIT $index, $quantity ";
+		}
+
+		if ( main::DEBUGLOG && $sqllog->is_debug ) {
+			$sqllog->debug( "Roles query: $sql / " . Data::Dump::dump($p) );
+		}
+
+		my $sth = $dbh->prepare_cached($sql);
+		$sth->execute( @{$p} );
+
+		$roles = $sth->fetchrow_array();
+		$sth->finish;
+
+		if ( !$stillScanning ) {
+			$cache->{$cacheKey} = $roles;
+		}
+	}
+
+	$request->addResult( 'role_ids', [ split(',',$roles) ] );
```

Comment:
Is this the same as 

```perl
foreach (@$roles) {
   $request->addResultLoop('role_ids', $chunkCount++, 'role_id', $_);
}
```

or similar? If not, please change. I'd really want this to be in line with all the other queries as much as possible. Or it'll byte us at some point...

Reply:
Decided to return a single array of roles, rather than the usual result_loop, in discussion with @CDrummond, but it does make sense for query results to be to some sort of standard, so I'll change to a loop with a value for each role. Also allows adding optional extra data, like contributor description as you suggested above.

Tagging Craig here so he knows what's going on.